In [54]:
import pandas as pd
import numpy as np
import datetime

In [2]:
#read book data csv into a df
books = pd.read_csv('orbitBooks_12132018.csv', index_col = 0)
books_clean = books.copy()

In [3]:
books_clean

title                author           isbn  \
0                    The Devil You Know            Mike Carey  9780446197175   
1                The Devil's Right Hand      Lilith Saintcrow  9780316021425   
2                 Working for the Devil      Lilith Saintcrow     0316003131   
3                     The Innocent Mage          Karen Miller  9780316067805   
4                   The Electric Church           Jeff Somers  9780316021722   
5                           Winterbirth         Brian Ruckley  9780316067690   
6                       Dead Man Rising      Lilith Saintcrow  9780316003148   
7                       Dead Man Rising      Lilith Saintcrow  9780316019552   
8                The Devil's Right Hand      Lilith Saintcrow  9780316019446   
9                 Working for the Devil      Lilith Saintcrow  9780316019491   
10                  The Electric Church           Jeff Somers  9780316019385   
11                  Devices and Desires          K. J. Parker     0316003387   
12               Once Bitten, Twice Shy       Jennifer Rardin  9780316020466   
13                    The Awakened Mage          Karen Miller     0316067814   
14               Once Bitten, Twice Shy       Jennifer Rardin  9780316022750   
15                    A Forest of Stars     Kevin J. Anderson     031600345X   
16                        Hidden Empire     Kevin J. Anderson     0316003441   
17                       Horizon Storms     Kevin J. Anderson     0316003476   
18                    Of Fire and Night     Kevin J. Anderson  9780316021739   
19                       Scattered Suns     Kevin J. Anderson     0316003484   
20                        Evil for Evil          K. J. Parker     0316003395   
21                   Saint City Sinners      Lilith Saintcrow  9780316021432   
22                   Saint City Sinners      Lilith Saintcrow  9780316023276   
23                          Metal Swarm     Kevin J. Anderson  9780316021746   
24           Another One Bites the Dust       Jennifer Rardin  9780316020572   
25                       The Escapement          K. J. Parker  9780316003407   
26               The Devil's Right Hand      Lilith Saintcrow  9780316021425   
27                          Metal Swarm     Kevin J. Anderson  9780316024044   
28           Another One Bites the Dust       Jennifer Rardin  9780316023979   
29                  The Ashes of Worlds     Kevin J. Anderson  9780316007573   
...                                 ...                   ...            ...   
1235                        Charmcaster  Sebastien de Castell  9780316525824   
1236                   Children of Time    Adrian Tchaikovsky  9780316452496   
1237                   The Game of Luck     Catherine Cerveny  9780316441667   
1238                  Persepolis Rising     James S. A. Corey  9780316332859   
1239           The Broken Earth Trilogy         N. K. Jemisin  9780316527194   
1240             There Before the Chaos          K. B. Wagers  9780316411219   
1241             There Before the Chaos          K. B. Wagers  9780316411202   
1242                   Treason of Hawks            Lila Bowen  9780316502375   
1243                   The Wastelanders         K. S. Merbeth  9780316527071   
1244                   Treason of Hawks            Lila Bowen  9780316502382   
1245                   The Wastelanders         K. S. Merbeth  9780316527033   
1246                          Cold Iron         Miles Cameron  9780316399319   
1247                           Red Moon  Kim Stanley Robinson  9780316262378   
1248                           Red Moon  Kim Stanley Robinson  9780316262354   
1249                          Cold Iron         Miles Cameron  9780316399326   
1250                   Season of Storms     Andrzej Sapkowski  9780316441629   
1251                    Someone Like Me           M. R. Carey  9780316477420   
1252                    Someone Like Me           M. R. Carey  9780316477444   
1253                    The Winter

## Check for duplicates and null values

In [4]:
books_clean.query('title == "Everything About You"')

Empty DataFrame
Columns: [title, author, isbn, cover_price, page_count, book_type, book_format, publication_date, publisher, cover_image]
Index: []

In [5]:
#check for duplicates
isbns = books_clean['isbn']
#code taken from: https://stackoverflow.com/questions/14657241/how-do-i-get-a-list-of-all-the-duplicate-items-using-pandas-in-python
books_clean[isbns.isin(isbns[isbns.duplicated()])].sort_values("isbn")

title                  author           isbn  \
29           The Ashes of Worlds       Kevin J. Anderson  9780316007573   
62           The Ashes of Worlds       Kevin J. Anderson  9780316007573   
1         The Devil's Right Hand        Lilith Saintcrow  9780316021425   
26        The Devil's Right Hand        Lilith Saintcrow  9780316021425   
193   Lord of the Changing Winds         Rachel Neumeier  9780316072786   
223   Lord of the Changing Winds         Rachel Neumeier  9780316072786   
617               Abaddon's Gate       James S. A. Corey  9780316129077   
680               Abaddon's Gate       James S. A. Corey  9780316129077   
567             Promise of Blood         Brian McClellan  9780316219044   
709             Promise of Blood         Brian McClellan  9780316219044   
1075               The Stone Sky           N. K. Jemisin  9780316229241   
1077               The Stone Sky           N. K. Jemisin  9780316229241   
904                The Immortals    Jordanna Max Brodsky  9780316347181   
911                The Immortals    Jordanna Max Brodsky  9780316347181   
856       Before They Are Hanged         Joe Abercrombie  9780316387354   
1198      Before They Are Hanged         Joe Abercrombie  9780316387354   
857       Last Argument of Kings         Joe Abercrombie  9780316387408   
1148      Last Argument of Kings         Joe Abercrombie  9780316387408   
1175              Blood of Elves       Andrzej Sapkowski  9780316438988   
1101              Blood of Elves       Andrzej Sapkowski  9780316438988   
267              The Company Man  Robert Jackson Bennett  9781841498621   
323              The Company Man  Robert Jackson Bennett  9781841498621   

     cover_price page_count book_type book_format publication_date  \
29        $25.99   xvii+491     NOVEL          hc       2008-07-00   
62        $25.99   xvii+491     NOVEL          hc       2008-07-00   
1          $6.99        391     NOVEL          pb       2007-09-00   
26         $6.99        391     NOVEL          pb       2007-09-00   
193        $7.99        367     NOVEL          pb       2010-05-00   
223        $7.99        367     NOVEL          pb       2010-05-00   
617       $17.00        566     NOVEL          tp       2013-06-00   
680       $17.00        566     NOVEL          tp       2013-06-00   
567       $16.00        545     NOVEL          tp       2014-04-08   
709       $16.00        545     NOVEL          tp       2014-04-08   
1075      $16.99        445     NOVEL          tp       2017-08-15   
1077      $16.99        445     NOVEL          tp       2017-08-15   
904       $25.00        449     NOVEL          hc       2016-02-16   
911       $25.00        449     NOVEL          hc       2016-02-16   
856       $16.99        514     NOVEL          tp       2015-09-08   
1198      $16.99        514     NOVEL          tp       2015-09-08   
857       $16.99        603     NOVEL          tp       2015-09-08   
1148      $16.99        603     NOVEL          tp       2015-09-08   
1175      $15.99        432     NOVEL          tp       2017-10-03   
1101      $15.99        432     NOVEL          tp       2017-10-03   
267          NaN        464     NOVEL          hc       2010-10-00   
323          NaN        464     NOVEL          hc       2010-10-00   

       publisher                                        cover_image  
29    Orbit (US)  http://www.isfdb.org/wiki/images/e/eb/THSHSFWR...  
62    Orbit (US)  http://www.isfdb.org/wiki/images/e/eb/THSHSFWR...  
1     Orbit (US)  http://ecx.images-amazon.com/images/I/51mYofAd...  
26    Orbit (US)  http://ecx.images-amazon.com/images/I/51mYofAd...  
193   Orbit (US)  http://ecx.images-amazon.com/images/I/51t2sgSE...  
223   Orbit (US)  http://ecx.images-amazon.com/images/I/51t2sgSE...  
617   Orbit (US)  http://www.isfdb.org/wiki/images/b/b5/BDDNSGTT...  
680   Orbit (US)  http://www.isfdb.org/wiki/images/b/b5/BDDNSGTT...  
567   Orbit (US)  http://ecx.images-amazon.com/images/I/51-G

In [6]:
#remove duplicates and null rows 
books_clean = books_clean.drop_duplicates()
books_clean = books_clean.dropna(how='all')
len(books_clean)

1253

In [7]:
books_clean = books_clean.reset_index(drop=True)

## Super General Assessment of Data

In [8]:
books_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1253 entries, 0 to 1252
Data columns (total 10 columns):
title               1253 non-null object
author              1253 non-null object
isbn                1253 non-null object
cover_price         1249 non-null object
page_count          665 non-null object
book_type           1253 non-null object
book_format         1253 non-null object
publication_date    1253 non-null object
publisher           1253 non-null object
cover_image         1243 non-null object
dtypes: object(10)
memory usage: 98.0+ KB


## Fix unknown book formats

In [9]:
books_clean.book_format.value_counts()

ebook      590
tp         416
pb         144
hc         101
unknown      2
Name: book_format, dtype: int64

Notes: <br>
hc is hardcover <br>
tp is a trade paperback <br>
pb is a mass market paperback <Br>
ebook is an ebook (file type not specified)<br>
<br>
There currently are no audiobooks listed in this dataset. That will need to be changed. 

In [10]:
books_clean.query('book_format == "unknown"')

title           author           isbn cover_price page_count  \
209  The Gaslight Dogs   Karin Lowachee  9780316021791       $7.99        368   
266    Bitten to Death  Jennifer Rardin  9780316075541       $7.99        320   

    book_type book_format publication_date   publisher  \
209     NOVEL     unknown       2010-03-30  Orbit (US)   
266     NOVEL     unknown       2010-10-01  Orbit (US)   

                                           cover_image  
209  http://ecx.images-amazon.com/images/I/51ZI17g7...  
266                                                NaN

In [11]:
books_clean.query('title == "The Gaslight Dogs"')

title          author           isbn cover_price page_count  \
209  The Gaslight Dogs  Karin Lowachee  9780316021791       $7.99        368   
213  The Gaslight Dogs  Karin Lowachee  9780316088473       $6.99        NaN   

    book_type book_format publication_date   publisher  \
209     NOVEL     unknown       2010-03-30  Orbit (US)   
213     NOVEL       ebook       2010-04-01  Orbit (US)   

                                           cover_image  
209  http://ecx.images-amazon.com/images/I/51ZI17g7...  
213  http://ecx.images-amazon.com/images/I/514Yg9GD...

9780316021791 is the pb version. It has a dash in the ISBN-13 on Amazon, which may be why it didn't pull that data.

In [12]:
books_clean.query('title == "Bitten to Death"')

title           author           isbn cover_price page_count  \
66   Bitten to Death  Jennifer Rardin  9780316022088      $12.99        301   
72   Bitten to Death  Jennifer Rardin  9780316032919       $9.99        NaN   
266  Bitten to Death  Jennifer Rardin  9780316075541       $7.99        320   

    book_type book_format publication_date   publisher  \
66      NOVEL          tp       2008-08-00  Orbit (US)   
72      NOVEL       ebook       2008-08-12  Orbit (US)   
266     NOVEL     unknown       2010-10-01  Orbit (US)   

                                           cover_image  
66   http://ecx.images-amazon.com/images/I/513TEG77...  
72   http://ecx.images-amazon.com/images/I/51ODUTGK...  
266                                                NaN

9780316075541 is the pb version. It has a dash in the ISBN-13 on Amazon, which may be why it didn't pull that data.

In [13]:
books_clean.book_format = books_clean.book_format.replace('unknown', 'pb')

In [14]:
books_clean.query('book_format == "unknown"')

Empty DataFrame
Columns: [title, author, isbn, cover_price, page_count, book_type, book_format, publication_date, publisher, cover_image]
Index: []

In [15]:
books_clean.query('title == "Bitten to Death"')

title           author           isbn cover_price page_count  \
66   Bitten to Death  Jennifer Rardin  9780316022088      $12.99        301   
72   Bitten to Death  Jennifer Rardin  9780316032919       $9.99        NaN   
266  Bitten to Death  Jennifer Rardin  9780316075541       $7.99        320   

    book_type book_format publication_date   publisher  \
66      NOVEL          tp       2008-08-00  Orbit (US)   
72      NOVEL       ebook       2008-08-12  Orbit (US)   
266     NOVEL          pb       2010-10-01  Orbit (US)   

                                           cover_image  
66   http://ecx.images-amazon.com/images/I/513TEG77...  
72   http://ecx.images-amazon.com/images/I/51ODUTGK...  
266                                                NaN

## Resolve price errors

Prices are missing from 5 books. Eventually the datatype in this column will need to be float, so all characters that are not integers or periods will need to be removed. Also, some of the prices are listed in British currency, while most are listed in US currency. All prices will need to be converted to US currency.

In [16]:
#create a dataframe that has cover prices
cover_price = books_clean.query("cover_price == cover_price")

In [17]:
cover_price[cover_price['cover_price'].str.contains("£")]

title           author           isbn cover_price page_count  \
175      Bite Marks  Jennifer Rardin  9780316043823       £9.99        336   
542   The Perimeter    Will McIntosh  9781405524872       £1.49        NaN   
1217       Magefall    Stephen Aryan  9780356508498       £9.99        438   

     book_type book_format publication_date   publisher  \
175      NOVEL          tp       2010-04-00       Orbit   
542   CHAPBOOK       ebook       2012-10-15  Orbit (US)   
1217     NOVEL          tp       2018-09-06       Orbit   

                                            cover_image  
175   http://ecx.images-amazon.com/images/I/51LmlTpg...  
542   http://ecx.images-amazon.com/images/I/51-Si9Rb...  
1217  https://images-na.ssl-images-amazon.com/images...

indices from book_cleans with irregular values:<br>
92 listed the price as $7.99 US <br>
179 UK cover price<br>
226 has no price<br>
267 has no price<br>
323 has no price<br>
393 lists no currency<br>
546 UK cover price<br>
717 is a row full of NaNs<br>
877 has no price<br>
1074 has no price<br>
1231 UK cover price<br>

In [18]:
cover_price[cover_price['cover_price'].str.contains("\x801.49")]

title             author           isbn  \
387  The Butcher of Anderson Station  James S. A. Corey  9780316204071   

    cover_price page_count book_type book_format publication_date   publisher  \
387       €1.49         40  CHAPBOOK       ebook       2011-10-17  Orbit (US)   

                                           cover_image  
387  http://ecx.images-amazon.com/images/I/51HKC4A7...

In [19]:
books_clean.iloc[1217]

title                                                        Magefall
author                                                  Stephen Aryan
isbn                                                    9780356508498
cover_price                                                     £9.99
page_count                                                        438
book_type                                                       NOVEL
book_format                                                        tp
publication_date                                           2018-09-06
publisher                                                       Orbit
cover_image         https://images-na.ssl-images-amazon.com/images...
Name: 1217, dtype: object

In [20]:
#drop book published in UK
books_clean = books_clean.drop(1217, axis=0)

In [21]:
books_clean.iloc[542]
#correct price is $1.99

title                                                   The Perimeter
author                                                  Will McIntosh
isbn                                                    9781405524872
cover_price                                                     £1.49
page_count                                                        NaN
book_type                                                    CHAPBOOK
book_format                                                     ebook
publication_date                                           2012-10-15
publisher                                                  Orbit (US)
cover_image         http://ecx.images-amazon.com/images/I/51-Si9Rb...
Name: 542, dtype: object

In [22]:
books_clean.iloc[175]

title                                                      Bite Marks
author                                                Jennifer Rardin
isbn                                                    9780316043823
cover_price                                                     £9.99
page_count                                                        336
book_type                                                       NOVEL
book_format                                                        tp
publication_date                                           2010-04-00
publisher                                                       Orbit
cover_image         http://ecx.images-amazon.com/images/I/51LmlTpg...
Name: 175, dtype: object

In [23]:
books_clean.query('title == "Magefall"')

title         author           isbn cover_price page_count book_type  \
1215  Magefall  Stephen Aryan  9780316554817      $16.99        438     NOVEL   
1216  Magefall  Stephen Aryan  9780356508481       $6.99        NaN     NOVEL   

     book_format publication_date   publisher  \
1215          tp       2018-09-04  Orbit (US)   
1216       ebook       2018-09-04  Orbit (US)   

                                            cover_image  
1215  https://images-na.ssl-images-amazon.com/images...  
1216  https://images-na.ssl-images-amazon.com/images...

In [24]:
#correct price of book published in US but showing in £
books_clean.cover_price = books_clean.cover_price.replace('£9.99', '$13.99')
books_clean.cover_price = books_clean.cover_price.replace('£1.49', '$1.99')
books_clean.cover_price = books_clean.cover_price.replace('\x801.49', '$2.99')
books_clean.publisher = books_clean.publisher.replace('Orbit', 'Orbit (US)')

In [25]:
books_clean.query('title == "Bite Marks"')

title           author           isbn cover_price page_count  \
169  Bite Marks  Jennifer Rardin  9780316071697       $9.99        NaN   
175  Bite Marks  Jennifer Rardin  9780316043823      $13.99        336   

    book_type book_format publication_date   publisher  \
169     NOVEL       ebook       2009-10-14  Orbit (US)   
175     NOVEL          tp       2010-04-00  Orbit (US)   

                                           cover_image  
169  http://ecx.images-amazon.com/images/I/51wTd8Bv...  
175  http://ecx.images-amazon.com/images/I/51LmlTpg...

In [26]:
#test that both books have been fixed
books_clean.query('publisher == "Orbit"')

Empty DataFrame
Columns: [title, author, isbn, cover_price, page_count, book_type, book_format, publication_date, publisher, cover_image]
Index: []

In [27]:
#check missing prices
books_clean.query('cover_price != cover_price')

title                  author           isbn cover_price  \
223            Shadowrise            Tad Williams  9781841492957         NaN   
264       The Company Man  Robert Jackson Bennett  9781841498621         NaN   
870       The Vital Abyss       James S. A. Corey  9780316402811         NaN   
1064  The Prince of Glass            Karen Miller  9780316120159         NaN   

     page_count book_type book_format publication_date   publisher  \
223         592     NOVEL          tp       2010-05-06  Orbit (US)   
264         464     NOVEL          hc       2010-10-00  Orbit (US)   
870         NaN  CHAPBOOK       ebook       2015-10-15  Orbit (US)   
1064        512     NOVEL          hc       2017-08-01  Orbit (US)   

                                            cover_image  
223   http://ecx.images-amazon.com/images/I/41CJHTfN...  
264   http://ecx.images-amazon.com/images/I/51jleTkn...  
870   https://images-na.ssl-images-amazon.com/images...  
1064                                                NaN

In [28]:
#Company Man list price is $34 (source: https://www.bookdepository.com/Company-Man-Robert-Jackson-Bennett/9781841498621)
books_clean['cover_price'][264] = '$34'

#Shadowrise list price is $20.55 (source: https://www.bookdepository.com/Shadowrise-Tad-Williams/9781841492957)
books_clean['cover_price'][223] = '$20.55'

#The Vital Abyss list price is $2.99, change ISBN to 9780316217569 (https://www.kobo.com/us/en/ebook/the-vital-abyss-1)
books_clean['cover_price'][870] = '$2.99'
books_clean['isbn'][870] = '9780316217569'

#The Prince of Glass isn't out yet and should be dropped
books_clean = books_clean.drop(1064, axis=0)

In [29]:
books_clean.query('cover_price != cover_price')

Empty DataFrame
Columns: [title, author, isbn, cover_price, page_count, book_type, book_format, publication_date, publisher, cover_image]
Index: []

In [30]:
books_clean.query('title=="Shadowrise"')

title        author           isbn cover_price page_count book_type  \
223  Shadowrise  Tad Williams  9781841492957      $20.55        592     NOVEL   

    book_format publication_date   publisher  \
223          tp       2010-05-06  Orbit (US)   

                                           cover_image  
223  http://ecx.images-amazon.com/images/I/41CJHTfN...

In [31]:
#drop $ for cover_price, then change column datatype to float
books_clean.cover_price = books_clean.cover_price.str.replace('[$US]', '').astype(float)

books_clean

title                author           isbn  \
0                    The Devil You Know            Mike Carey  9780446197175   
1                The Devil's Right Hand      Lilith Saintcrow  9780316021425   
2                 Working for the Devil      Lilith Saintcrow     0316003131   
3                     The Innocent Mage          Karen Miller  9780316067805   
4                   The Electric Church           Jeff Somers  9780316021722   
5                           Winterbirth         Brian Ruckley  9780316067690   
6                       Dead Man Rising      Lilith Saintcrow  9780316003148   
7                       Dead Man Rising      Lilith Saintcrow  9780316019552   
8                The Devil's Right Hand      Lilith Saintcrow  9780316019446   
9                 Working for the Devil      Lilith Saintcrow  9780316019491   
10                  The Electric Church           Jeff Somers  9780316019385   
11                  Devices and Desires          K. J. Parker     0316003387   
12               Once Bitten, Twice Shy       Jennifer Rardin  9780316020466   
13                    The Awakened Mage          Karen Miller     0316067814   
14               Once Bitten, Twice Shy       Jennifer Rardin  9780316022750   
15                    A Forest of Stars     Kevin J. Anderson     031600345X   
16                        Hidden Empire     Kevin J. Anderson     0316003441   
17                       Horizon Storms     Kevin J. Anderson     0316003476   
18                    Of Fire and Night     Kevin J. Anderson  9780316021739   
19                       Scattered Suns     Kevin J. Anderson     0316003484   
20                        Evil for Evil          K. J. Parker     0316003395   
21                   Saint City Sinners      Lilith Saintcrow  9780316021432   
22                   Saint City Sinners      Lilith Saintcrow  9780316023276   
23                          Metal Swarm     Kevin J. Anderson  9780316021746   
24           Another One Bites the Dust       Jennifer Rardin  9780316020572   
25                       The Escapement          K. J. Parker  9780316003407   
26                          Metal Swarm     Kevin J. Anderson  9780316024044   
27           Another One Bites the Dust       Jennifer Rardin  9780316023979   
28                  The Ashes of Worlds     Kevin J. Anderson  9780316007573   
29                      Debatable Space         Philip Palmer  9780316018920   
...                                 ...                   ...            ...   
1223                        Charmcaster  Sebastien de Castell  9780316525824   
1224                   Children of Time    Adrian Tchaikovsky  9780316452496   
1225                   The Game of Luck     Catherine Cerveny  9780316441667   
1226                  Persepolis Rising     James S. A. Corey  9780316332859   
1227           The Broken Earth Trilogy         N. K. Jemisin  9780316527194   
1228             There Before the Chaos          K. B. Wagers  9780316411219   
1229             There Before the Chaos          K. B. Wagers  9780316411202   
1230                   Treason of Hawks            Lila Bowen  9780316502375   
1231                   The Wastelanders         K. S. Merbeth  9780316527071   
1232                   Treason of Hawks            Lila Bowen  9780316502382   
1233                   The Wastelanders         K. S. Merbeth  9780316527033   
1234                          Cold Iron         Miles Cameron  9780316399319   
1235                           Red Moon  Kim Stanley Robinson  9780316262378   
1236                           Red Moon  Kim Stanley Robinson  9780316262354   
1237                          Cold Iron         Miles Cameron  9780316399326   
1238                   Season of Storms     Andrzej Sapkowski  9780316441629   
1239                    Someone Like Me           M. R. Carey  9780316477420   
1240                    Someone Like Me           M. R. Carey  9780316477444   
1241                    The Winter

## Clean up page counts

Page count is missing from about half -- however since ebooks are part of the sample, missing the page count may mean that book doesn't have pages because of the format. We can get around that by adding a word count column so that we can at least tell the approximate length of the books. The word count will be calculated by multiplying 350 and the book's page count. This number was calculated based on assertions made here: https://www.writersservices.com/writersservices-self-publishing/word-count-page (350 given as the average word count on a page). 

In [32]:
books_clean.page_count.value_counts()

432         9
464         8
320         8
390         7
416         6
370         6
384         6
593         6
496         5
326         5
373         5
553         5
438         5
344         5
486         5
374         5
309         5
336         4
434         4
436         4
419         4
455         4
576         4
369         4
512         4
402         4
400         4
353         4
516         4
359         4
           ..
447         1
vi+350      1
506         1
399         1
609         1
302         1
485         1
295         1
xvii+691    1
519         1
514         1
[8]+516     1
348         1
300         1
387         1
390+[13]    1
735         1
530         1
1216        1
465         1
572         1
360         1
672         1
285         1
xx+756      1
1248        1
784         1
321         1
875         1
655         1
Name: page_count, Length: 374, dtype: int64

In [33]:
books_clean.iloc[546][4]
#these need to be added together

'373+388+355+374+386'

In [34]:
#fix the page count value for the Parasol Protectorate Boxed Set
books_clean.page_count = books_clean.page_count.replace('373+388+355+374+386', '1876')

In [36]:
pgCt_booksClean = books_clean.query('page_count==page_count')

In [37]:
pgCt_booksClean['page_count'] =  pgCt_booksClean.page_count.str.replace('[a-z/W+]', '')
pgCt_booksClean['page_count'] =  pgCt_booksClean.page_count.str.replace('\[[0-9]\]|\[[0-9][0-9]\]', '')
pgCt_booksClean['page_count']

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


1        391
2        384
3        642
4        353
5        541
6        380
11       635
12       291
13       712
15       669
16       639
17       626
18       659
19       691
20       684
21       371
23       461
24       319
25       409
28       491
29       488
30       389
31       701
33       593
34       315
36       654
38       527
39       405
40       424
41       665
        ... 
1209     315
1212    1242
1213     510
1215     438
1218     476
1220     398
1221     390
1222     369
1225     430
1226     592
1227    1424
1228     480
1230     384
1231     608
1234     445
1235     464
1238     464
1239     512
1241     496
1242     496
1243    1249
1244     528
1245     416
1246     432
1247     688
1248     528
1249     640
1250     896
1251     544
1252     416
Name: page_count, Length: 663, dtype: object

In [38]:
pgCt_booksClean['page_count'] = pgCt_booksClean.page_count.astype(int)
pgCt_booksClean.page_count.dtype

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


dtype('int64')

## Add word count column

In [39]:
pgCt_booksClean.drop_duplicates(subset='title', inplace=True)
pgCt_booksClean

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


title                author  \
1                         The Devil's Right Hand      Lilith Saintcrow   
2                          Working for the Devil      Lilith Saintcrow   
3                              The Innocent Mage          Karen Miller   
4                            The Electric Church           Jeff Somers   
5                                    Winterbirth         Brian Ruckley   
6                                Dead Man Rising      Lilith Saintcrow   
11                           Devices and Desires          K. J. Parker   
12                        Once Bitten, Twice Shy       Jennifer Rardin   
13                             The Awakened Mage          Karen Miller   
15                             A Forest of Stars     Kevin J. Anderson   
16                                 Hidden Empire     Kevin J. Anderson   
17                                Horizon Storms     Kevin J. Anderson   
18                             Of Fire and Night     Kevin J. Anderson   
19                                Scattered Suns     Kevin J. Anderson   
20                                 Evil for Evil          K. J. Parker   
21                            Saint City Sinners      Lilith Saintcrow   
23                                   Metal Swarm     Kevin J. Anderson   
24                    Another One Bites the Dust       Jennifer Rardin   
25                                The Escapement          K. J. Parker   
28                           The Ashes of Worlds     Kevin J. Anderson   
29                               Debatable Space         Philip Palmer   
30                              To Hell and Back      Lilith Saintcrow   
31                  Across the Face of the World   Russell Kirkpatrick   
33                                        Matter         Iain M. Banks   
34                             Biting the Bullet       Jennifer Rardin   
36                 In the Earth Abides the Flame   Russell Kirkpatrick   
38                              Consider Phlebas         Iain M. Banks   
39                           The Player of Games         Iain M. Banks   
40                                   Black Ships             Jo Graham   
41                         The Right Hand of God   Russell Kirkpatrick   
...                                          ...                   ...   
1200                           King of Assassins          R. J. Barker   
1201                               Friendly Fire            Dale Lucas   
1202               The Tower of Living and Dying      Anna Smith Spark   
1207                                   Bad Faith           Jon Hollins   
1208                           Blood of the Gods         David Mealing   
1209                                 Shadowblack  Sebastien de Castell   
1212  The Collected Adventures of Bannon & Clare      Lilith Saintcrow   
1213                                 Bloody Rose        Nicholas Eames   
1215                                    Magefall         Stephen Aryan   
1218                    The Land You Never Leave          Angus Watson   
1221                                   Rosewater         Tade Thompson   
1222                                 Charmcaster  Sebastien de Castell   
1225                            The Game of Luck     Catherine Cerveny   
1227                    The Broken Earth Trilogy         N. K. Jemisin   
1228                      There Before the Chaos          K. B. Wagers   
1230                            Treason of Hawks            Lila Bowen   
1231                            The Wastelanders         K. S. Merbeth   
1234                                   Cold Iron         Miles Cameron   
1235                                    Red Moon  Kim Stanley Robinson   
1239                             Someone Like Me           M. R. Carey   
1241                             The Winter Road          Adrian Selby   
1242                              Empire of Sand            Tasha Suri   
1244                                      Exodu

In [40]:
#calculate word_count column
avg_words_pp = 350
pgCt_booksClean.loc[:,'page_count'] *= avg_words_pp
#rename page_count to word_count; drop all columns except title and word_count
pgCt_booksClean = pgCt_booksClean.drop(columns = ['author', 'isbn', 'cover_price', 'book_type', 'book_format', 'publication_date', 'publisher','cover_image'])
pgCt_booksClean.rename(index=str, inplace=True, columns={'page_count': 'approximate_word_count'})
pgCt_booksClean

/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:537: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


title  approximate_word_count
1                         The Devil's Right Hand                  136850
2                          Working for the Devil                  134400
3                              The Innocent Mage                  224700
4                            The Electric Church                  123550
5                                    Winterbirth                  189350
6                                Dead Man Rising                  133000
11                           Devices and Desires                  222250
12                        Once Bitten, Twice Shy                  101850
13                             The Awakened Mage                  249200
15                             A Forest of Stars                  234150
16                                 Hidden Empire                  223650
17                                Horizon Storms                  219100
18                             Of Fire and Night                  230650
19                                Scattered Suns                  241850
20                                 Evil for Evil                  239400
21                            Saint City Sinners                  129850
23                                   Metal Swarm                  161350
24                    Another One Bites the Dust                  111650
25                                The Escapement                  143150
28                           The Ashes of Worlds                  171850
29                               Debatable Space                  170800
30                              To Hell and Back                  136150
31                  Across the Face of the World                  245350
33                                        Matter                  207550
34                             Biting the Bullet                  110250
36                 In the Earth Abides the Flame                  228900
38                              Consider Phlebas                  184450
39                           The Player of Games                  141750
40                                   Black Ships                  148400
41                         The Right Hand of God                  232750
...                                          ...                     ...
1200                           King of Assassins                  176400
1201                               Friendly Fire                  151900
1202               The Tower of Living and Dying                  178850
1207                                   Bad Faith                  177450
1208                           Blood of the Gods                  240800
1209                                 Shadowblack                  110250
1212  The Collected Adventures of Bannon & Clare                  434700
1213                                 Bloody Rose                  178500
1215                                    Magefall                  153300
1218                    The Land You Never Leave                  166600
1221                                   Rosewater                  136500
1222                                 Charmcaster                  129150
1225                            The Game of Luck                  150500
1227                    The Broken Earth Trilogy                  498400
1228                      There Before the Chaos                  168000
1230                            Treason of Hawks                  134400
1231                            The Wastelanders                  212800
1234                                   Cold Iron                  155750
1235                                    Red Moon                  162400
1239                             Someone Like Me                  179200
1241                             The Winter Road                  173600
1242                              Empire of Sand                  173600
1244                                      Exodus                  184800
1245           How Long 'til

In [41]:
#merge df w/ word count column back into main df on title
books_clean = books_clean.merge(pgCt_booksClean, how='outer', on='title')
books_clean

title                author           isbn  \
0                    The Devil You Know            Mike Carey  9780446197175   
1                    The Devil You Know            Mike Carey  9780316511766   
2                The Devil's Right Hand      Lilith Saintcrow  9780316021425   
3                The Devil's Right Hand      Lilith Saintcrow  9780316019446   
4                 Working for the Devil      Lilith Saintcrow     0316003131   
5                 Working for the Devil      Lilith Saintcrow  9780316019491   
6                     The Innocent Mage          Karen Miller  9780316067805   
7                     The Innocent Mage          Karen Miller  9780316212793   
8                   The Electric Church           Jeff Somers  9780316021722   
9                   The Electric Church           Jeff Somers  9780316019385   
10                  The Electric Church           Jeff Somers  9780316053938   
11                          Winterbirth         Brian Ruckley  9780316067690   
12                          Winterbirth         Brian Ruckley  9780316068062   
13                          Winterbirth         Brian Ruckley  9780316068314   
14                      Dead Man Rising      Lilith Saintcrow  9780316003148   
15                      Dead Man Rising      Lilith Saintcrow  9780316019552   
16                  Devices and Desires          K. J. Parker     0316003387   
17                  Devices and Desires          K. J. Parker  9780316077477   
18               Once Bitten, Twice Shy       Jennifer Rardin  9780316020466   
19               Once Bitten, Twice Shy       Jennifer Rardin  9780316022750   
20               Once Bitten, Twice Shy       Jennifer Rardin  9780316043540   
21                    The Awakened Mage          Karen Miller     0316067814   
22                    The Awakened Mage          Karen Miller  9780316212779   
23                    A Forest of Stars     Kevin J. Anderson     031600345X   
24                        Hidden Empire     Kevin J. Anderson     0316003441   
25                       Horizon Storms     Kevin J. Anderson     0316003476   
26                    Of Fire and Night     Kevin J. Anderson  9780316021739   
27                       Scattered Suns     Kevin J. Anderson     0316003484   
28                        Evil for Evil          K. J. Parker     0316003395   
29                        Evil for Evil          K. J. Parker  9780316077484   
...                                 ...                   ...            ...   
1221                           Magefall         Stephen Aryan  9780356508481   
1222           The Land You Never Leave          Angus Watson  9780316317399   
1223           The Land You Never Leave          Angus Watson  9780316317405   
1224                        Charmcaster  Sebastien de Castell  9780316525848   
1225                        Charmcaster  Sebastien de Castell  9780316525824   
1226                   Children of Time    Adrian Tchaikovsky  9780316452496   
1227                   Children of Time    Adrian Tchaikovsky  9780316452502   
1228                   The Game of Luck     Catherine Cerveny  9780316441667   
1229           The Broken Earth Trilogy         N. K. Jemisin  9780316527194   
1230             There Before the Chaos          K. B. Wagers  9780316411219   
1231             There Before the Chaos          K. B. Wagers  9780316411202   
1232                   Treason of Hawks            Lila Bowen  9780316502375   
1233                   Treason of Hawks            Lila Bowen  9780316502382   
1234                   The Wastelanders         K. S. Merbeth  9780316527071   
1235                   The Wastelanders         K. S. Merbeth  9780316527033   
1236                          Cold Iron         Miles Cameron  9780316399319   
1237                          Cold Iron         Miles Cameron  9780316399326   
1238                           Red Moon  Kim Stanley Robinson  9780316262378   
1239                           Red

## Resolve publication date issues & change publication date to datetime object

publication_date needs to be a datetime object. page_count needs to be integers. cover_price needs to be a float. book_type and book_format should probably both be category datatypes.

In [42]:
books_clean.publication_date.value_counts()

2012-09-04    27
2017-11-07     9
2011-07-01     9
2017-12-05     9
2017-03-14     8
2007-11-00     7
2011-08-01     7
2012-10-30     7
2016-09-06     7
2012-05-01     7
2013-05-07     6
2007-09-00     6
2011-01-01     6
2017-09-26     6
2018-08-07     6
2017-06-27     6
2017-11-14     6
2016-12-20     6
2014-04-01     5
2018-09-18     5
2018-08-21     5
2015-07-28     5
2011-06-00     5
2016-04-26     5
2018-07-17     5
2017-10-31     5
2016-12-06     5
2016-07-19     5
2017-08-22     5
2018-06-05     5
              ..
2008-02-11     1
2008-08-12     1
2009-09-04     1
2009-05-01     1
2012-11-00     1
2012-11-29     1
2014-04-29     1
2014-02-04     1
2016-07-05     1
2007-07-10     1
2009-07-03     1
2009-09-24     1
2010-04-29     1
2013-09-03     1
2011-04-07     1
2011-10-07     1
2011-05-12     1
2015-09-01     1
2016-11-29     1
2011-08-17     1
2011-12-15     1
2018-02-06     1
2016-01-05     1
2010-02-01     1
2011-09-11     1
2018-09-25     1
2014-08-05     1
2008-12-00    

The dates listed sometimes end in 00, so to correct that, for now I'm dropping the date information.

In [43]:
books_clean[books_clean['publication_date'].str.contains("-00")]
#there's 161 with bad dates that need to be fixed

title                author           isbn  \
2         The Devil's Right Hand      Lilith Saintcrow  9780316021425   
4          Working for the Devil      Lilith Saintcrow     0316003131   
6              The Innocent Mage          Karen Miller  9780316067805   
8            The Electric Church           Jeff Somers  9780316021722   
10           The Electric Church           Jeff Somers  9780316053938   
11                   Winterbirth         Brian Ruckley  9780316067690   
12                   Winterbirth         Brian Ruckley  9780316068062   
14               Dead Man Rising      Lilith Saintcrow  9780316003148   
16           Devices and Desires          K. J. Parker     0316003387   
18        Once Bitten, Twice Shy       Jennifer Rardin  9780316020466   
20        Once Bitten, Twice Shy       Jennifer Rardin  9780316043540   
23             A Forest of Stars     Kevin J. Anderson     031600345X   
24                 Hidden Empire     Kevin J. Anderson     0316003441   
25                Horizon Storms     Kevin J. Anderson     0316003476   
26             Of Fire and Night     Kevin J. Anderson  9780316021739   
27                Scattered Suns     Kevin J. Anderson     0316003484   
28                 Evil for Evil          K. J. Parker     0316003395   
30            Saint City Sinners      Lilith Saintcrow  9780316021432   
32                   Metal Swarm     Kevin J. Anderson  9780316021746   
33                   Metal Swarm     Kevin J. Anderson  9780316024044   
34                   Metal Swarm     Kevin J. Anderson  9780316021753   
35    Another One Bites the Dust       Jennifer Rardin  9780316020572   
37    Another One Bites the Dust       Jennifer Rardin  9780316043557   
38                The Escapement          K. J. Parker  9780316003407   
40           The Ashes of Worlds     Kevin J. Anderson  9780316007573   
42           The Ashes of Worlds     Kevin J. Anderson  9780316007580   
43               Debatable Space         Philip Palmer  9780316018920   
44               Debatable Space         Philip Palmer  9780316068093   
46              To Hell and Back      Lilith Saintcrow  9780316001779   
49                        Matter         Iain M. Banks  9780316005364   
...                          ...                   ...            ...   
479                         2312  Kim Stanley Robinson  9780316098120   
483             The King's Blood        Daniel Abraham  9780316080774   
496               Tempest's Fury         Nicole Peeler  9780316128117   
498                Caliban's War     James S. A. Corey  9780316129060   
520      The Fate of the Dwarves          Markus Heitz  9780316102629   
522         The Iron Wyrm Affair      Lilith Saintcrow  9780316201261   
529               Seeds of Earth        Michael Cobley  9780316213981   
530               Seeds of Earth        Michael Cobley  9780316213998   
564                  Red Country       Joe Abercrombie  9780316187213   
567          The Orphaned Worlds        Michael Cobley  9780316214018   
568          The Orphaned Worlds        Michael Cobley  9780316214001   
569          The Hydrogen Sonata         Iain M. Banks  9780316212373   
570          The Hydrogen Sonata         Iain M. Banks  9780316212380   
571          The Hydrogen Sonata         Iain M. Banks  9780316212366   
581          The Ascendant Stars        Michael Cobley  9780316214032   
589         The Damnation Affair      Lilith Saintcrow  9780316251594   
626        The Red Plague Affair      Lilith Saintcrow  9780316183734   
640               Abaddon's Gate     James S. A. Corey  9780316129077   
659                       Shaman  Kim Stanley Robinson  9780316098076   
700                  Dirty Magic            Jaye Wells  9780316228435   
780                     War Dogs             Greg Bear  9780316072830   
855                       Aurora  Kim Stanley Robinson  9780316098106   
856                       Aurora  Kim Stanley Robinson  9780316378741   
857         

In [44]:
#When I have time I'm going to fix the -00 publication dates.
#books_clean['publication_date'][2] = '2007-09-01'
#books_clean['publication_date'][4] = '2007-09-25'
#books_clean['publication_date'][6] = '2007-09-01'
#books_clean['publication_date'][8] =  
#books_clean['publication_date'][10] =  
#books_clean['publication_date'][11] =  
#books_clean['publication_date'][12] =  
#books_clean['publication_date'][14] =  
#books_clean['publication_date'][15] =  
#books_clean['publication_date'][16] =   
#books_clean['publication_date'][18] =  
#books_clean['publication_date'][19] =  
#books_clean['publication_date'][20] =  
#books_clean['publication_date'][23] =  
#books_clean['publication_date'][24] =  
#books_clean['publication_date'][25] =  
#books_clean['publication_date'][26] =  
#books_clean['publication_date'][27] = 
#books_clean['publication_date'][28] =  
#books_clean['publication_date'][30] =  
#books_clean['publication_date'][32] =  
#books_clean['publication_date'][33] =  
#books_clean['publication_date'][34] =  
#books_clean['publication_date'][35] =  
#books_clean['publication_date'][37] =  
#books_clean['publication_date'][38] =  
#books_clean['publication_date'][40] =  
#books_clean['publication_date'][42] =  
#books_clean['publication_date'][43] =  
#books_clean['publication_date'][44] =  
#books_clean['publication_date'][46] =   
#books_clean['publication_date'][49] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] = 
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][] =  
#books_clean['publication_date'][479] =  
#books_clean['publication_date'][474] =  
#books_clean['publication_date'][475] =  
#books_clean['publication_date'][497] =  
#books_clean['publication_date'][498] =  
#books_clean['publication_date'][506] =  
#books_clean['publication_date'][507] =  
#books_clean['publication_date'][541] =  
#books_clean['publication_date'][542] =  
#books_clean['publication_date'][543] =  
#books_clean['publication_date'][544] =  
#books_clean['publication_date'][545] =  
#books_clean['publication_date'][555] =  
#books_clean['publication_date'][578] =  
#books_clean['publication_date'][603] =  
#books_clean['publication_date'][617] =  
#books_clean['publication_date'][626] =  
#books_clean['publication_date'][640] =  
#books_clean['publication_date'][641] =  
#books_clean['publication_date'][681] =  
#books_clean['publication_date'][763] =  
#books_clean['publication_date'][834] =  
#books_clean['publication_date'][835] =  
#books_clean['publication_date'][836] =  
#books_clean['publication_date'][867] =  
#books_clean['publication_date'][868] =  
#books_clean['publication_date'][978] =  
#books_clean['publication_date'][1147] =  
#books_clean['publication_date'][1149] =  
#books_clean['publication_date'][1150] =  

In [45]:
books_clean.publication_date = books_clean.publication_date.str.replace('-00', '')

In [46]:
books_clean[books_clean['publication_date'].str.contains("-00")]

Empty DataFrame
Columns: [title, author, isbn, cover_price, page_count, book_type, book_format, publication_date, publisher, cover_image, approximate_word_count]
Index: []

In [47]:
books_clean[books_clean['publication_date'].str.contains("date unknown")]

title        author        isbn  cover_price page_count  \
21   The Awakened Mage  Karen Miller  0316067814         7.99        712   
113      Shadow's Edge   Brent Weeks  0316033650         7.99    [6]+645   

    book_type book_format publication_date   publisher  \
21      NOVEL          pb     date unknown  Orbit (US)   
113     NOVEL          pb     date unknown  Orbit (US)   

                                           cover_image  approximate_word_count  
21   http://ecx.images-amazon.com/images/I/51G%2BFG...                249200.0  
113  http://www.isfdb.org/wiki/images/7/71/SHDWSDGJ...                225750.0

In [48]:
books_clean['publication_date'][21] =  '2007-10-01'
books_clean['publication_date'][113] =  '2008-11-01'

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [55]:
books_clean['publication_date'] = pd.to_datetime(books_clean['publication_date'], format= '%Y-%m') 

In [ ]:
books_clean['publication_year'] = pd.to_datetime(books_clean['publication_date'], format='%Y-%m').dt.year

In [56]:
books_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1251 entries, 0 to 1250
Data columns (total 11 columns):
title                     1251 non-null object
author                    1251 non-null object
isbn                      1251 non-null object
cover_price               1251 non-null float64
page_count                663 non-null object
book_type                 1251 non-null object
book_format               1251 non-null object
publication_date          1251 non-null datetime64[ns]
publisher                 1251 non-null object
cover_image               1242 non-null object
approximate_word_count    1131 non-null float64
dtypes: datetime64[ns](1), float64(2), object(8)
memory usage: 157.3+ KB


In [51]:
#this list will eventually be used to get award data information
#create a dataframe that only has unique titles
#title = books_clean.drop(columns=['cover_image', 'publisher', 'cover_price', 'book_format', 'book_type', 'publication_date', 'page_count'])
#title.drop_duplicates('title', inplace=True)

#send the unique titles to a CSV
#title.to_csv('titles.csv')

In [53]:
books_clean.to_csv('books.csv')